### PR TITLE
Feature/112 reconfigure the detectors to use global code for the binary serializers 2

### PR DESCRIPTION
**Pull request by hayakawa16**
Creating a draft PR so it goes through all of the checks.  Will add reviewers next week.

### CODE REVIEW

**Comment by lmalenfant** [comment]
@hayakawa16 The quality gate passed!

There are 11 new code smells, we can't fix them all (some are cognitive complexity with the factory code) but some of them can be fixed.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs [hunk 1]
```diff
@@ -404,263 +404,40 @@ public void Normalize(long numPhotons)
...
+                    DynamicMTOfZ, "DynamicMTOfZ","_DynamicMTOfZ"),
+                BinaryArraySerializerFactory.GetSerializer(
+                    SubregionCollisions,"SubregionCollisions","_SubregionCollisions"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    TotalMTOfZSecondMoment, "TotalMTOfZSecondMoment", "_TotalMTOfZ_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    DynamicMTOfZSecondMoment, "DynamicMTOfZSecondMoment", "_DynamicMTOfZ_2") : null
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:437]: Remove extra line.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs [hunk 2]
```diff
@@ -404,263 +404,40 @@ public void Normalize(long numPhotons)
...
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:440]: Remove extra line.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfRhoAndSubregionHistDetector.cs [hunk 3]
```diff
@@ -356,219 +356,39 @@ public void Normalize(long numPhotons)
...
+                    SubregionCollisions,"SubregionCollisions","_SubregionCollisions"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    TotalMTOfZSecondMoment, "TotalMTOfZSecondMoment", "_TotalMTOfZ_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    DynamicMTOfZSecondMoment, "DynamicMTOfZSecondMoment", "_DynamicMTOfZ_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfRhoAndSubregionHistDetector.cs:391]: Remove extra line.

Add a line before the comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetector.cs [hunk 4]
```diff
@@ -378,286 +378,39 @@ public void Normalize(long numPhotons)
...
+                    SubregionCollisions, "SubregionCollisions", "_SubregionCollisions"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                        SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                        TotalMTOfZSecondMoment, "TotalMTOfZSecondMoment", "_TotalMTOfZ_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                        DynamicMTOfZSecondMoment, "DynamicMTOfZSecondMoment", "_DynamicMTOfZ_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetector.cs:413]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs [hunk 5]
```diff
@@ -260,85 +261,24 @@ public void Normalize(long numPhotons)
...
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    FractionalMT, "FractionalMT", "_FractionalMT"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs:281]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs [hunk 6]
```diff
@@ -270,97 +271,24 @@ public void Normalize(long numPhotons)
...
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    FractionalMT, "FractionalMT", "_FractionalMT"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs:291]: Remove extra line.

Add line above comments.

File: src/Vts/MonteCarlo/Detectors/dMCdROfRhodMuaDetector.cs [hunk 7]
```diff
@@ -268,43 +268,20 @@ public void Normalize(long numPhotons)
...
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
         /// <summary>
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/dMCdROfRhodMuaDetector.cs:286]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/dMCdROfRhodMusDetector.cs [hunk 8]
```diff
@@ -293,43 +293,20 @@ public void Normalize(long numPhotons)
...
+
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/dMCdROfRhodMusDetector.cs:310]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/pMCROfFxAndTimeDetector.cs [hunk 9]
```diff
@@ -227,61 +227,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfFxAndTimeDetector.cs:243]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/pMCROfFxDetector.cs [hunk 10]
```diff
@@ -210,49 +210,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfFxDetector.cs:226]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeDetector.cs [hunk 11]
```diff
@@ -210,55 +210,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeDetector.cs:226]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeRecessedDetector.cs [hunk 12]
```diff
@@ -224,55 +224,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfRhoAndTimeRecessedDetector.cs:240]: Add a line.

File: src/Vts/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetector.cs [hunk 13]
```diff
@@ -266,111 +266,26 @@ public void Normalize(long numPhotons)
...
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    ROfXAndY, "ROfXAndY", "_ROfXAndY"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    ROfXAndYSecondMoment, "ROfXAndYSecondMoment", "_ROfXAndY_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetector.cs:288]: Remove line.

Add line above the comments.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeDetector.cs [hunk 14]
```diff
@@ -188,63 +190,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeDetector.cs:206]: Add a line above the comments.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetectorTests.cs [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+                    { { 7, 8, 9 }, { 10, 11, 12 } }
+            },
+            SecondMoment = new double[,,]
+            {
+                    { { 13, 14, 15 }, { 16, 17, 18 } },
+                    { { 19, 20, 21 }, { 22, 23, 24 } }
+            },
+            FractionalMT = new double[,,,,] // X.Count-1 x Y.Count-1 x MTBins.Count-1 x NumRegions x FractionalMTBins.Count+1: 2x2x3x2x2
+            {
+                    {
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetectorTests.cs:54]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/SurfaceFiberTests.cs
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/SurfaceFiberTests.cs:1]: Another good catch.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetectorTests.cs [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        {
+            Fx = new DoubleRange(0, 10, 2),
+            MTBins = new DoubleRange(0, 10, 4),
+            Z = new DoubleRange(0, 10, 4),
+            FractionalMTBins = new DoubleRange(0, 1, 1),
+            TallySecondMoment = tallySecondMoment,
+            Name = detectorName,
+            Mean = new[,] // Fx.Count x MTBins.Count-1: 2x3
+            {
+                    { 1 + Complex.ImaginaryOne, 2 + 2 * Complex.ImaginaryOne, 3 + 3 * Complex.ImaginaryOne },
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetectorTests.cs:45]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+            X = new DoubleRange(-10, 10, 3),
+            Y = new DoubleRange(-10, 10, 3),
+            MTBins = new DoubleRange(0, 100, 4),
+            Z = new DoubleRange(0, 100, 4),
+            FractionalMTBins = new DoubleRange(0, 10, 1),
+            TallySecondMoment = tallySecondMoment,
+            Name = detectorName,
+            Mean = new double[,,] // X.Count-1 x Y.Count-1 x MTBins.Count-1: 2x2x3
+            {
+                    { { 1, 2, 3 }, { 4, 5, 6 } },
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:45]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetectorTests.cs [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+                    { { 7, 8, 9 }, { 10, 11, 12 } }
+            },
+            SecondMoment = new double[,,]
+            {
+                    { { 13, 14, 15 }, { 16, 17, 18 } },
+                    { { 19, 20, 21 }, { 22, 23, 24 } }
+            },
+            FractionalMT = new double[,,,,] // X.Count-1 x Y.Count-1 x MTBins.Count-1 x NumRegions x FractionalMTBins.Count+1: 2x2x3x2x2
+            {
+                    {
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetectorTests.cs:54]: Fixed

File: src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfFxAndTimeTests.cs
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfFxAndTimeTests.cs:1]: Fixed.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetectorTests.cs [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+            X = new DoubleRange(-10, 10, 3),
+            Y = new DoubleRange(-10, 10, 3),
+            Time = new DoubleRange(0, 1, 3),
+            TallySecondMoment = tallySecondMoment,
+            PerturbedOps = new List<OpticalProperties> { new() },
+            PerturbedRegionsIndices = new List<int> { 1 },
+            Name = detectorName,
+            Mean = new double[,,,] // X.Count-1 x Z.Count-1 x Time.Count-1 x NumOfRegions: 2x2x2x3
+            {
+                    {
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts.Test/Unit/MonteCarlo/Detectors/pMCROfXAndYAndTimeAndSubregionDetectorTests.cs:46]: Fixed

File: src/Vts/MonteCarlo/Detectors/ATotalDetector.cs [hunk 20]
```diff
@@ -1,5 +1,6 @@
 using System;
 using System.Linq;
+using System.Reflection.Metadata.Ecma335;
```
[the remark this answers is not in the document]
Comment (hayakawa16) [reply src/Vts/MonteCarlo/Detectors/ATotalDetector.cs:3]: Done.

**Review by hayakawa16 — COMMENTED** [review]
Thank you for your very detailed review!  I made code changes for all suggestions, retested and pushed.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetectorTests.cs [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+        Assert.AreEqual(15 + 15 * Complex.ImaginaryOne, detector.TotalMTOfZ[0, 2]);
+        Assert.AreEqual(16 + 16 * Complex.ImaginaryOne, detector.TotalMTOfZ[1, 0]);
+        Assert.AreEqual(17 + 17 * Complex.ImaginaryOne, detector.TotalMTOfZ[1, 1]);
+        Assert.AreEqual(18 + 18 * Complex.ImaginaryOne, detector.TotalMTOfZ[1, 2]);
+        Assert.AreEqual(25 + 25 * Complex.ImaginaryOne, detector.DynamicMTOfZ[0, 0]);
+        Assert.AreEqual(26 + 26 * Complex.ImaginaryOne, detector.DynamicMTOfZ[0, 1]);
+        Assert.AreEqual(27 + 27 * Complex.ImaginaryOne, detector.DynamicMTOfZ[0, 2]);
+        Assert.AreEqual(28 + 28 * Complex.ImaginaryOne, detector.DynamicMTOfZ[1, 0]);
+        Assert.AreEqual(29 + 29 * Complex.ImaginaryOne, detector.DynamicMTOfZ[1, 1]);
+        Assert.AreEqual(30 + 30 * Complex.ImaginaryOne, detector.DynamicMTOfZ[1, 2]); Assert.AreEqual(37 + 37 * Complex.ImaginaryOne, detector.FractionalMT[0, 0, 0]);
```
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetectorTests.cs:111]: 2 statements on one line.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+
+                },
+                {
+                    { { 13, 14 }, { 15, 16 }, { 17, 18 } },
+                    { { 19, 20 }, { 21, 22 }, { 23, 24 } }
+                }
+            },
+            SubregionCollisions = new double[,] // NumSubregions x 2: 3x2
+                { { 25, 26 }, { 27, 28 }, { 29, 30 } }, // 2nd index: 0=static, 1=dynamic
+
```
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/ReflectedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:86]: Extra line.

File: src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs [hunk 23]
```diff
@@ -0,0 +1,202 @@
...
+                    { { 7, 8 }, { 9, 10 }, { 11, 12 } }
+                },
+                {
+                    { { 13, 14 }, { 15, 16 }, { 17, 18 } },
+                    { { 19, 20 }, { 21, 22 }, { 23, 24 } }
+                }
+            },
+            SubregionCollisions = new double[,] // NumSubregions x 2: 3x2
+                { { 25, 26 }, { 27, 28 }, { 29, 30 } }, // 2nd index: 0=static, 1=dynamic
+
```
Comment (lmalenfant) [inline src/Vts.Test/Unit/MonteCarlo/Detectors/TransmittedDynamicMTOfXAndYAndSubregionHistDetectorTests.cs:85]: Extra line.

File: src/Vts/MonteCarlo/Detectors/ROfRhoAndOmegaDetector.cs [hunk 24]
```diff
@@ -182,61 +183,20 @@ public void Normalize(long numPhotons)
...
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
+
         }
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ROfRhoAndOmegaDetector.cs:200]: Add line above comments.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndMaxDepthDetector.cs [hunk 25]
```diff
@@ -188,63 +190,19 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
[the remark this answers is not in the document]
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ROfXAndYAndMaxDepthDetector.cs:205]: Line is still missing here.

File: src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeAndSubregionDetector.cs [hunk 26]
```diff
@@ -228,116 +230,25 @@ public void Normalize(long numPhotons)
...
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    ROfXAndY, "ROfXAndY", "_ROfXAndY"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    ROfXAndYSecondMoment, "ROfXAndYSecondMoment", "_ROfXAndY_2") : null
             };
+            return allSerializers.Where(s => s is not null).ToArray();
         }
```
[the remark this answers is not in the document]
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ROfXAndYAndTimeAndSubregionDetector.cs:251]: Line is still missing.

File: src/Vts/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetector.cs [hunk 27]
```diff
@@ -381,238 +381,40 @@ public void Normalize(long numPhotons)
...
+                    DynamicMTOfZ, "DynamicMTOfZ","_DynamicMTOfZ"),
+                BinaryArraySerializerFactory.GetSerializer(
+                    SubregionCollisions,"SubregionCollisions","_SubregionCollisions"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    TotalMTOfZSecondMoment, "TotalMTOfZSecondMoment", "_TotalMTOfZ_2") : null,
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    DynamicMTOfZSecondMoment, "DynamicMTOfZSecondMoment", "_DynamicMTOfZ_2") : null
+
```
[the remark this answers is not in the document]
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/ReflectedDynamicMTOfFxAndSubregionHistDetector.cs:415]: Having a line before the closing bracket of an object definition could result in a maintainer putting code in the wrong place.

File: src/Vts/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetector.cs [hunk 28]
```diff
@@ -267,104 +268,33 @@ public void Normalize(long numPhotons)
...
+
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    FractionalMT, "FractionalMT", "_FractionalMT"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/ReflectedMTOfXAndYAndSubregionHistDetector.cs:293]: Line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs (hunk 1 above)
Comment (lmalenfant) [reply src/Vts/MonteCarlo/Detectors/TransmittedDynamicMTOfFxAndSubregionHistDetector.cs:434]: Extra line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs [hunk 29]
```diff
@@ -254,92 +254,32 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    FractionalMT, "FractionalMT", "_FractionalMT"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfRhoAndSubregionHistDetector.cs:278]: Extra line before the closing bracket.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs [hunk 30]
```diff
@@ -264,104 +265,32 @@ public void Normalize(long numPhotons)
...
+            }
+            var allSerializers = new List<BinaryArraySerializer>
+            {
+                BinaryArraySerializerFactory.GetSerializer(
+                    Mean, "Mean", ""),
+                BinaryArraySerializerFactory.GetSerializer(
+                    FractionalMT, "FractionalMT", "_FractionalMT"),
+                TallySecondMoment ? BinaryArraySerializerFactory.GetSerializer(
+                    SecondMoment, "SecondMoment", "_2") : null,
+
```
Comment (lmalenfant) [inline src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs:289]: Extra line before the closing bracket.

**Review by lmalenfant — CHANGES_REQUESTED** [review]
I have added a few more comments where there is a line feed before the end of an object definition, it is not intentional whitespace and I think it could result in a developer putting future code in the wrong place.

File: src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs (hunk 30 above)
Comment (hayakawa16) [reply src/Vts/MonteCarlo/Detectors/TransmittedMTOfXAndYAndSubregionHistDetector.cs]: All review comments implemented.  Thank you for all of your help!

**Comment by hayakawa16** [comment]
Thanks Lisa!  I will await dcuccia's review before proceeding.

**Comment by hayakawa16** [comment]
Just wanted to post that I took an infile that I ran with master on 10/12/23 and reran it using this branch.  The results are identical!  It was a sanity check suggested by lmalenfant and gives us more confidence in our code modifications.

**Review by dcuccia — APPROVED** [review]
Looks great - thanks for all the hard work that went into this. My one comment that we should consider changing before merging is that we update the names of the test methods. 'Validate_deserialized_class_is_correct_when_using_GetBinarySerializers` is less accurate than `Validate_deserialized_binary_arrays_are_correct_when_using_GetBinarySerializers`.

My comments regarding things we might consider for the future are:

1. That we refactor the test projects into separate Integration and Unit projects (and as part of that, generally better align test names to the classes/class members they're testing).
2. That we pare down `Vts.Test.MonteCarlo.DetectorIOTests` significantly, now that all detectors' serializers' behavior/correctness are already covered by these new unit tests (don't need duplicate checks - wastes time and increases software maintenance load). Remaining work would probably be to refactor what remains to unit tests that test the detector-specific JSON serialization/deserialization for the non-array properties.
3. Similarly, to refactor test classes in the `Vts.Test.MonteCarlo.Detectors` into their Unit counterparts, if feasible - that way we can declare + test the intended behavior of the specific detector without relying on executing a Monte Carlo simulation (i.e. correct tally accumulation and normalization, e.g. focus on correctness of code, not accuracy of MC).
4. Once 1 & 2 are done, we would be left with tests that test the accuracy of the Monte Carlo for those detectors, which would be probably best labeled as integration tests. (An implementation detail, but a potential way to do that rapidly be to use an in-memory post-processor database, so you don't have to "run" an MC for each detector, just walk through the photon data points).

Perhaps we could create a couple follow-on placeholder tickets (to work on when appropriate) that documents our desire to do the above?

**Comment by hayakawa16** [comment]
dcuccia, thank you for your review!  I can easily modify the unit test names. 

I like your ideas for future work.  It always bothered me how many times a MC simulation is executed during the course of running the unit tests.  If this could be optimized somehow, that would be great!  In addition to your list, in our discussion on this issue we uncovered the need to analyze PopulateFromEnumerable and possibly update application to be column-major.  Creating place holder issues for these features sounds good.  Would you like to create them?

**Pull request merged**